tiatoolbox/models/segmentation/semantic_segmentor.py

```diff
@@ -0,0 +1,983 @@
+                self._model,
+                sample_datas,
+                self._on_gpu,
+            )
+            # repackage so that its a N list, each contains
+            # L x etc. output
+            sample_outputs = [np.split(v, batch_size, axis=0) for v in sample_outputs]
+            sample_outputs = list(zip(*sample_outputs))
+
+            # tensor to numpy, costly?
+            sample_infos = sample_infos.numpy()
+            sample_infos = np.split(sample_infos, batch_size, axis=0)
+
+            sample_outputs = list(zip(sample_infos, sample_outputs))
+            cum_output.extend(sample_outputs)
+            pbar.update()
+        pbar.close()
+        self._process_predictions(cum_output, wsi_reader, ioconfig, save_path)
+
+    def _process_predictions(
+        self, cum_batch_predictions, wsi_reader, ioconfig, save_path
+    ):
+        """Define how the aggregated predictions are processed.
+
+        This includes merging the prediction if necessary and also saving afterward.
+
+        """
+        # assume predictions is N, each item has L output element
+        locations, predictions = list(zip(*cum_batch_predictions))
+        # Nx4 (N x [tl_x, tl_y, br_x, br_y), denotes the location of output patch
+        # this can exceed the image bound at the requested resolution
+        # remove singleton due to split.
+        locations = np.array([v[0] for v in locations])
+        for idx, output_resolution in enumerate(ioconfig.output_resolutions):
+            # assume resolution idx to be in the same order as L
+            merged_resolution = ioconfig.highest_input_resolution
+            merged_locations = locations
+            # ! location is wrt highest resolution, hence still need conversion
+            if ioconfig.save_resolution is not None:
+                merged_resolution = ioconfig.save_resolution
+                output_shape = wsi_reader.slide_dimensions(**output_resolution)
+                merged_shape = wsi_reader.slide_dimensions(**merged_resolution)
+                fx = merged_shape[0] / output_shape[0]
+                merged_locations = np.ceil(locations * fx).astype(np.int64)
+            merged_shape = wsi_reader.slide_dimensions(**merged_resolution)
+            # 0 idx is to remove singleton wihout removing other axes singleton
+            to_merge_predictions = [v[idx][0] for v in predictions]
+            sub_save_path = f"{save_path}.raw.{idx}.npy"
+            self.merge_prediction(
+                merged_shape[::-1],  # XY to YX
+                to_merge_predictions,
+                merged_locations,
+                save_path=sub_save_path,
+                free_prediction=True,
+            )
+
+    @staticmethod
+    def merge_prediction(
+        canvas_shape: Union[Tuple[int], List[int], np.ndarray],
+        predictions: List[np.ndarray],
+        locations: Union[List, np.ndarray],
+        save_path: Union[str, pathlib.Path] = None,
+        free_prediction: bool = True,
```

Comment:
Would 'in_place` be a better name for this arg?

Reply:
`in_place` implies some operation works on the item itself, and the items technically still exist. This function uses the items to construct another output and the items staying the same. `free_prediction` is asking if the user allows discarding items during such construction.